File: snippets/csharp/session-management/marker-management.cs
// ─────────────────────────────────────────────────────────────
// SQL Race Example: Marker Management
// ─────────────────────────────────────────────────────────────
//
// Demonstrates: Creating Marker objects with time ranges and MarkerLabel annotations
// Prerequisites: MESL.SQLRace.API NuGet package, .NET 8
// Input: None (creates its own session)
// Output: Markers with labels printed to console
//
// Related: snippets/csharp/session-management/lap-and-segment-management.cs
// Docs: https://mat-docs.github.io/Atlas.SQLRaceAPI.Documentation/api/index.html
// ─────────────────────────────────────────────────────────────

using MAT.OCS.Core;
using MESL.SqlRace.Domain;

Core.LicenceProgramName = "SQLRace";
Core.Initialize();

var connectionString = $"DbEngine=SQLite;Data Source={Path.Combine(Path.GetTempPath(), "sqlrace-examples.ssn2")};";
var sessionKey = SessionKey.NewKey();

var sessionManager = SessionManager.CreateSessionManager();
using var clientSession = sessionManager.CreateSession(connectionString, sessionKey, "Marker Demo", DateTime.UtcNow, "Example");
var session = clientSession.Session;

var baseTime = 36_000_000_000_000L; // 10:00:00.000

// --- Add markers representing annotated time regions ---
var anomaly = new Marker(baseTime, baseTime + 2_000_000_000L, "Anomaly", "Anomaly", "Unexpected temperature spike detected", Guid.NewGuid().ToString());
var roi = new Marker(baseTime + 5_000_000_000L, baseTime + 8_000_000_000L, "RegionOfInterest", "RegionOfInterest", "Steady-state operating window", Guid.NewGuid().ToString());
var transient = new Marker(baseTime + 10_000_000_000L, baseTime + 12_000_000_000L, "Transient", "Transient", "Load step response period", Guid.NewGuid().ToString());

session.Markers.Add(anomaly);
session.Markers.Add(roi);
session.Markers.Add(transient);

Console.WriteLine($"Session {sessionKey}: {session.Markers.Count} markers added\n");

foreach (var marker in session.Markers)
{
    var startNs = marker.StartTimestamp ?? 0;
    var endNs = marker.EndTimestamp ?? 0;
    var durationS = (endNs - startNs) / 1_000_000_000.0;
    Console.WriteLine($"  [{marker.Label}] {startNs}–{endNs} ns ({durationS:F1}s)");
    Console.WriteLine($"    Type: {marker.MarkerType}");
    Console.WriteLine($"    Description: {marker.Description}");
    Console.WriteLine();
}

File: snippets/matlab/session-management/marker_management.m
% ─────────────────────────────────────────────────────────────
% SQL Race Example: Marker Management
% ─────────────────────────────────────────────────────────────
%
% Demonstrates: Creating Marker objects with time ranges and annotations
% Prerequisites: MATLAB R2023a+, .NET 8 runtime, MESL.SqlRace.Domain DLL
%                Run dotnetenv('core') once before using this snippet.
% Input: None (creates its own .ssn2 file in the temp folder)
% Output: Markers with labels and durations printed to the command window
%
% Related: snippets/matlab/session-management/lap_segment_management.m
% Docs: https://atlas.motionapplied.com/developer-resources/atlas/sql-race/examples/session-loading/
% ─────────────────────────────────────────────────────────────

% --- Configure .NET runtime (run once per MATLAB session) ---
% dotnetenv('core');  % uncomment if not already set

% --- Load assemblies and initialise ---
sqlraceDll = getenv('SQLRACE_DLL_PATH');
if isempty(sqlraceDll)
    sqlraceDll = 'C:\Program Files\McLaren Applied Technologies\ATLAS 10\MESL.SqlRace.Domain.dll';
end
installDir = fileparts(sqlraceDll);
sysEventsDll = fullfile(installDir, 'Microsoft.Win32.SystemEvents.dll');
if exist(sysEventsDll, 'file')
    NET.addAssembly(sysEventsDll);
end
NET.addAssembly(sqlraceDll);
NET.addAssembly(fullfile(installDir, 'MAT.OCS.Core.dll'));
import MESL.SqlRace.Domain.*;
import MAT.OCS.Core.*;

NET.setStaticProperty('MESL.SqlRace.Domain.Core.LicenceProgramName', 'SQLRace');
Core.Initialize();

% --- Connection string (writable temp file) ---
dbPath = fullfile(tempdir, 'sqlrace-matlab-markers.ssn2');
if exist(dbPath, 'file'); delete(dbPath); end
connectionString = sprintf('DbEngine=SQLite;Data Source=%s;', dbPath);
sessionKey = SessionKey.NewKey();

sessionManager = SessionManager.CreateSessionManager();
clientSession = sessionManager.CreateSession(connectionString, sessionKey, ...
    'Marker Demo', System.DateTime.UtcNow, 'Example');

try
    session = clientSession.Session;
    baseTime = int64(36e12);  % 10:00:00.000

    % --- Add markers representing annotated time regions ---
    % 6-argument ctor: (startTimestamp, endTimestamp, label, markerType, description, guid)
    anomaly = Marker(baseTime, baseTime + int64(2e9), ...
        'Anomaly', 'Anomaly', 'Unexpected temperature spike detected', char(System.Guid.NewGuid().ToString()));
    roi = Marker(baseTime + int64(5e9), baseTime + int64(8e9), ...
        'RegionOfInterest', 'RegionOfInterest', 'Steady-state operating window', char(System.Guid.NewGuid().ToString()));
    transient = Marker(baseTime + int64(10e9), baseTime + int64(12e9), ...
        'Transient', 'Transient', 'Load step response period', char(System.Guid.NewGuid().ToString()));

    session.Markers.Add(anomaly);
    session.Markers.Add(roi);
    session.Markers.Add(transient);

    fprintf('Session %s: %d markers added\n\n', ...
        char(sessionKey.ToString()), session.Markers.Count);

    for i = 1:session.Markers.Count
        marker = session.Markers.Item(i - 1);
        startNs = int64(0); endNs = int64(0);
        if marker.StartTimestamp.HasValue; startNs = marker.StartTimestamp.Value; end
        if marker.EndTimestamp.HasValue;   endNs   = marker.EndTimestamp.Value;   end
        durationS = double(endNs - startNs) / 1e9;
        fprintf('  [%s] %d-%d ns (%.1fs)\n', char(marker.Label), startNs, endNs, durationS);
        fprintf('    Type: %s\n', char(marker.MarkerType));
        fprintf('    Description: %s\n\n', char(marker.Description));
    end
catch ex
    fprintf('Error: %s\n', ex.message);
end

clientSession.Dispose();
disp('Done.');
